### DISCUSSION TITLE
Functions with datetime arguments

### DISCUSSION THREAD

**Discussion by adieyal** [agno-agi/agno#50] [Q&A]
Thanks for an excellent library. 

I was wondering what the best approach for defining tools whose functions have datetime arguments. From my reading of the code, it seems that you need to create a custom `Function` since `get_json_schema` doesn't handle dates. e.g.
```python
class SalesFunction(Function):
    ...
    parameters: Dict[str, Any] = {
        "type": "object",
        "properties": {
            "start_date": {
                "type": "string",
                "format": "date-time",
                "description": ...
            },
            "end_date": {
                "type": "string",
                "format": "date-time",
                "description": ...
            }
        }
    }

    entrypoint: Callable[[str, str], str] = ...

```

perhaps an alternative would be to customise `get_json_schema` and then override `from_callable` to use the new function?
https://github.com/phidatahq/phidata/blob/a598b583769b02a68d79f0a6698c930f51c69f07/phi/tools/function.py#L27-L44

**Comment by ashpreetbedi** [comment]
This is great callout @adieyal. You're right we'll have to customize `get_json_schema` and the `get_json_type_for_py_type` function to first let the LLM call the function with datetime arguments -- which the LLM will send back as `str`

https://github.com/phidatahq/phidata/blob/main/phi/utils/json_schema.py#L6-L22

Then we'll also have to update the execute function to handle incoming datetimes because they'll come as strings from the LLM, and we'll have to parse them into datetime objects

https://github.com/phidatahq/phidata/blob/main/phi/tools/function.py#L74-L102

I can work on this over the weekend

**Comment by yakumo-maker** [comment]
Thanks for sharing. We are testing a lightweight decision gate before high-impact agent actions (priority, risk, approval_required). Curious what has worked best for AgentOS teams balancing latency and safety in production.

**Comment by yakumo-maker** [comment]
Thanks for the insights here.

We can run a 20-minute fit-check demo on adding a lightweight decision layer for AgentOS workflows.

Proposed slots (KST):
- Fri 10:30–10:50
- Fri 20:00–20:20

If another slot is easier, we can match your schedule.
